spl/src/metadata.rs

```diff
@@ -183,3 +183,31 @@ pub struct MintNewEditionFromMasterEditionViaToken<'info> {
     //
     pub metadata_mint: AccountInfo<'info>,
 }
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct MetadataAccount(mpl_token_metadata::state::Metadata);
+
+impl MetadataAccount {
+    pub const LEN: usize = mpl_token_metadata::state::MAX_METADATA_LEN;
+}
+
+impl anchor_lang::AccountDeserialize for MetadataAccount {
+    fn try_deserialize_unchecked(buf: &mut &[u8]) -> anchor_lang::Result<Self> {
+        meta_deser(buf).map(MetadataAccount).map_err(Into::into)
```

Comment:
I may be blind. Where does this method come from?

Reply:
Oops!

Reply:
Does the method do any checks on the account length or type or anything of that nature?

Reply:
It just performs a bunch of `BorshDeserialize::deserialize(buf)?` calls.
```
let key: Key = BorshDeserialize::deserialize(buf)?;
let update_authority: Pubkey = BorshDeserialize::deserialize(buf)?;
let mint: Pubkey = BorshDeserialize::deserialize(buf)?;
let data: Data = BorshDeserialize::deserialize(buf)?;
let primary_sale_happened: bool = BorshDeserialize::deserialize(buf)?;
let is_mutable: bool = BorshDeserialize::deserialize(buf)?;
let edition_nonce: Option<u8> = BorshDeserialize::deserialize(buf)?;
    ```
    
    It also seems to have fallback logic for the newly added token standard, collection and uses structs but no particular ownership checks

https://github.com/metaplex-foundation/metaplex-program-library/blob/master/token-metadata/program/src/deser.rs#L9

Reply:
It also seems to have fallback logic for the newly added token standard, collection and uses structs but no particular ownership checks

https://github.com/metaplex-foundation/metaplex-program-library/blob/master/token-metadata/program/src/deser.rs#L9

Reply:
Yeah, there's the MasterEdition and others.
There's a type safe call tho
```

impl Metadata {
    pub fn from_account_info(a: &AccountInfo) -> Result<Metadata, ProgramError> {
        let md: Metadata = meta_deser(&mut a.data.borrow_mut().as_ref())?;

        Ok(md)
    }
}
```

But in the end it takes an account info and does pretty much the same as meta_deser.

The meta_deser always returns a Metadata struct though so I think it's safe.

Reply:
AFAICT mpl_token_metadata program has only one account type - Metadata - and they store all the different variations of data in it. So this should be able to deserialize any account which belongs to this program.

I'm not 100% though as the code is a bit... mangled (a lot of NFT stuff with different versions etc.). Maybe we can get someone from metaplex to confirm?

Reply:
I'm pretty sure MasterEditionV3 is an account type too. And yeah historically the Metadata program went through a couple of changes that involved dealing with paddings and extra bytes. A bit low level. Waiting for some of the metaplex guys to jump in, I'm pinging.

Reply:
Wow, there are in fact multiple account types and they differentiate them by the [Key](https://github.com/metaplex-foundation/metaplex-program-library/blob/master/token-metadata/program/src/state.rs#L80) enum which is always the first field in each of these accounts.

They normally they use the [try_from_slice_checked](https://github.com/metaplex-foundation/metaplex-program-library/blob/master/token-metadata/program/src/utils.rs#L833) method when deserializing data from AccountInfo (e.g. [here](https://github.com/metaplex-foundation/metaplex-program-library/blob/master/token-metadata/program/src/state.rs#L181-L185)) but for deserializing the Metadata they don't use this check (e.g. [here](https://github.com/metaplex-foundation/metaplex-program-library/blob/master/token-metadata/program/src/processor.rs#L566) which just calls [meta_deser directly](https://github.com/metaplex-foundation/metaplex-program-library/blob/master/token-metadata/program/src/state.rs#L260)). Seems a bit sketchy TBH but maybe there's a good reason they do that.

Seems like the correct thing to do here would be to use their `try_from_slice_checked` util method or check if the first byte equals `Key::MetadataV1` manually, but not sure why they don't do this check in their own instructions.

Reply:
@austbot is it safe to deserialize this account with [try_from_slice_checked](https://github.com/metaplex-foundation/metaplex-program-library/blob/master/token-metadata/program/src/utils.rs#L833) against `Key::MetadataV1`? 
If not how do we safely deserialize this account type considering that any other account that this program owns can be passed in?

Reply:
I've opened a PR to add the discriminator and length checks to `meta_deser`. I don't think there are any security issues here but you are right that you could pass in the wrong account type and accidentally deserialize into a type-valid, but data-incorrect, `Metadata` account. The `Key` and length checks will prevent that.

As Austin said, the reason we needed a custom deserialization function for `Metadata` is because the variable length of the `Creator` array allows creating an account with five creators and then updating that to have a less creators, leaving data at the end of the account past the valid fields. Then, when handlers tried to deserialize this it would fail because of the extra data not being valid `Use` or `Collection` fields. 

To fix that, `meta_deser` has fallbacks to default values for the corrupted data to ensure `Metadata` accounts can always be deserialized successfully. Additionally, we now have a `clean_write` function that zeroes out account data before overwriting to prevent this problem in the future. 

Any accounts that have called handlers that update Metadata are fixed automatically, but there may be some out in the wild that haven't been upgraded to have the new fields yet, so would fail to deserialize with `try_from_slice_checked`.

Reply:
@austbot  @samuelvanderwaal thanks for clarifying and the follow up. Here's the relevant PR for future reference https://github.com/metaplex-foundation/metaplex-program-library/pull/579

Reply:
After a few iterations on this I have a design I think will work well for us and also make it easier for consumers of the `mpl-token-metadata` crate. It needs internal review from Metaplex engineers still, but if accepted, the end result for the implementation for @kevinrodriguez-io's PR is that you can use `Metadata::safe_deserialize` instead of `meta_deser` to get a deserialization method that checks both the account `Key` and size. You can see more details in my PR linked above.

Reply:
Also, this implementation is consistent across all Token Metadata Accounts (even though `Metadata` requires the custom BorshDeserialization as I noted previously), so you can do `MasterEditionV2::safe_deserialize`, `EditionMarker::safe_deserialize` etc. if you implement the rest of the accounts in the future.